src/BuiltInTools/dotnet-watch/HotReload/StaticFileHandler.cs

```diff
@@ -2,50 +2,88 @@
-            if (!file.IsStaticFile || server is null)
+                if (file.StaticWebAssetPath is null)
+                {
+                    allFilesHandled = false;
+                    continue;
+                }
+
+                reporter.Verbose($"Handling file change event for static content {file.FilePath}.");
+
+                foreach (var containingProjectPath in file.ContainingProjectPaths)
+                {
+                    if (!projectMap.Map.TryGetValue(containingProjectPath, out var projectNodes))
+                    {
+                        // Shouldn't happen.
+                        reporter.Warn($"Project '{containingProjectPath}' not found in the project graph.");
+                        return allFilesHandled;
+                    }
+
+                    foreach (var projectNode in projectNodes)
+                    {
+                        if (browserConnector.TryGetRefreshServer(projectNode, out var refreshServer))
+                        {
+                            if (!refreshRequests.TryGetValue(refreshServer, out var filesPerServer))
+                            {
+                                refreshRequests.Add(refreshServer, filesPerServer = []);
+                            }
+
+                            filesPerServer.Add(file.StaticWebAssetPath);
+                        }
+                    }
+                }
+            }
+
+            if (refreshRequests.Count == 0)
             {
-                HotReloadEventSource.Log.HotReloadEnd(HotReloadEventSource.StartType.StaticHandler);
-                return false;
+                return allFilesHandled;
             }
-            _reporter.Verbose($"Handling file change event for static content {file.FilePath}.");
-            await HandleBrowserRefresh(server, file, cancellationToken);
-            HotReloadEventSource.Log.HotReloadEnd(HotReloadEventSource.StartType.StaticHandler);
-            _reporter.Output("Hot reload of static file succeeded.", emoji: "🔥");
-            return true;
-        }
 
-        private static async Task HandleBrowserRefresh(BrowserRefreshServer browserRefreshServer, FileItem fileItem, CancellationToken cancellationToken)
-        {
-            var message = JsonSerializer.SerializeToUtf8Bytes(new UpdateStaticFileMessage { Path = fileItem.StaticWebAssetPath }, JsonSerializerOptions);
-            await browserRefreshServer.SendMessage(message, cancellationToken);
+            var tasks = refreshRequests.Select(async request =>
+            {
+                // Serialize all requests sent to a single server:
+                foreach (var path in request.Value)
+                {
+                    reporter.Verbose($"Sending static file update request for asset '{path}'");
+                    var message = JsonSerializer.SerializeToUtf8Bytes(new UpdateStaticFileMessage { Path = path }, s_jsonSerializerOptions);
```

Comment:
I know this is the same as before, but is there a link to the receiver code that expects UTF8?

Reply:
Yes, no changes in the serialization

Reply:
The receiver is [in [WebSocketScriptInjection.js]](https://github.com/dotnet/sdk/blob/main/src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js#L46)